torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,863 @@
+from __future__ import annotations
+
+import dataclasses
+import functools
+import itertools
+import threading
+import warnings
+import weakref
+from collections import defaultdict
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+)
+from weakref import ref as WeakRef
+
+import torch.fx
+from torch import Tensor
+
+from torch._C import _cuda_CUDAAllocator_AllocatorState as AllocatorState
+from torch._dynamo.mutation_guard import GenerationTracker
+
+from torch._inductor.compile_fx import (
+    get_expanded_dims,
+    index_expanded_dims,
+    remove_unaligned_input_idxs,
+    static_input,
+)
+from torch._prims_common import check
+
+from torch.utils import _pytree as pytree
+from . import config
+
+
+@dataclasses.dataclass(frozen=True)
+class GraphID:
+    "Unique counter of a cuda graph recording"
+    id: int
+
+
+@dataclasses.dataclass(frozen=True)
+class FunctionID:
+    "Unique counter of a function wrapped in cudagraphify_impl"
+    id: int
+
+
+@dataclasses.dataclass(frozen=True)
+class WrappedFunction(object):
+    model: Callable
+    static_input_idxs: Sequence[int]
+    id: FunctionID
+
+
+class TreeManagerContainer(object):
+    """
+    Manages the lifetime of the tree manager. Like `PrivatePool` in cuda caching allocator,
+    the tree and its corresponding memory pool should be kept alive as long as any outstanding
+    graph or tensor which is an output of a graph remains alive.
```

Comment:
An explanation of the lifecycle of this object would be very useful. Based on reading, it seems like it is something like:

1. Is constructed, no graph, no fns, no tensors
2. Tree manager is fetched, resulting in tree manager being allocated
3. We generate a bunch of functions, calling add_strong_reference
4. These functions die, calling finalize_reference
5. When all the functions die, we finalize_tree_manager. We look for all the live tensors and add references to THOSE
6. We count as tensors die
7. All the tensors are dead, we deallocate the tree manager

Something that is not immediately obvious from reading this class only is the step 2/3 of the life cycle: what if you start building the tree manager but the first fn you make strong reference to dies. Won't you die prematurely? What if I then try to add another strong reference after having already finalized?

Reply:
> what if you start building the tree manager but the first fn you make strong reference to dies. Won't you die prematurely?

If the reference is dead, how is it dying prematurely. That is dying at the right time, no ? We only die when all references to the manager is gone. If there are multiple references and only one refernece dies, it wont be cleaned up

Reply:
Added this comment inline